### PR TITLE
Ea/story9to11

### PR DESCRIPTION
**Pull request by eamouzou**
- Create favoriteslist poro
- Add test, route, and view for favorite indicator in topnav
- Move flash message to application view
- Add favorite pet link to pets show page
- Add route, controller action, and redirect for favorite pet link
- Create test, route, controller action, and view for favorites index page

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,3 +1,12 @@
...
+
+  helper_method :favorites
+
+  def favorites
+    # ||= is saying set session[:favorites] to Hash.new(0) if it is nil, false, or undefined
+    # Otherwise set it to session[:favorites] meaning do nothing
+    # Memoization
+    @favorites ||= FavoritesList.new(session[:favorites])
+  end
 end
```
Comment (eamouzou) [inline app/controllers/application_controller.rb]: favorites helper method memoized here to be used in our favorites controller

File: app/controllers/favorites_controller.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  def update
+    pet = Pet.find(params[:pet_id])
+    favorites.add_pet(pet.id)
+    session[:favorites] = favorites.pets_favorite
+    quantity = favorites.count_of(pet.id)
+    flash[:favorite] = "The pet has been added to your favorites list."
+    redirect_to "/pets/#{pet.id}"
+  end
```
Comment (eamouzou) [inline app/controllers/favorites_controller.rb]: Methods being called on favorites are in the FavoritesList PORO.

File: app/controllers/favorites_controller.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class FavoritesController < ApplicationController
+  include ActionView::Helpers::TextHelper
+
+  def index
+    @fav_pets = favorites.pets_favorite.keys.map do |petid|
+      Pet.find(petid)
+    end
```
Comment (eamouzou) [inline app/controllers/favorites_controller.rb]: Might be an opportunity for refactoring here; feel free to adjust as you see fit.

File: app/models/favorites_list.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    @pets_favorite.values.sum
+  end
+
+  def add_pet(pet_id)
+    @pets_favorite[pet_id.to_s] = count_of(pet_id) + 1
+  end
+
+  def count_of(pet_id)
+    @pets_favorite[pet_id.to_s].to_i
+  end
```
Comment (eamouzou) [inline app/models/favorites_list.rb]: count_of allows it to turn a nil value to a count of 0.

File: app/views/favorites/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+
+<% @fav_pets.each do |pet| %>
+  <section id="favorite<%= pet.id %>">
+    <%= image_tag(pet.image, {:style => "width:200px;"}) %>
```
Comment (eamouzou) [inline app/views/favorites/index.html.erb]: O.pportunity to refactor: image style is showing up weird in server. Look it over when you get a chance.

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -11,10 +11,16 @@
...
   </div>
 
-
   <body>
+    <% flash.each do |type, msg| %>
+      <section class=<%= type %>>
+        <p><font size="5" color="red"><%= msg %></font></p>
+      </section>
+    <% end %>
+
```
Comment (eamouzou) [inline app/views/layouts/application.html.erb]: moved flash to application view

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -11,10 +11,16 @@
   <div class="topnav">
     <%=link_to "All Pets", "/pets"%>
     <%=link_to "All Shelters", "/shelters"%>
+    <%=link_to "Favorites: #{pluralize(favorites.total_count, "pet")}", '/favorites' %>
```
Comment (eamouzou) [inline app/views/layouts/application.html.erb]: keeps count of favorited pets, but ran into weird bug where it did not or could not reset even after reset

File: app/views/favorites/index.html.erb (hunk 5 above)
Comment (Will-Meighan) [reply app/views/favorites/index.html.erb]: Will do.

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (Will-Meighan) [reply app/views/layouts/application.html.erb]: Like this decision. Removing excess code from all other views.

**Comment by Will-Meighan** [comment]
Great work! Going to start on user stories 12 - 15.

**Pull request merged**